File: jscomp/test/record_extension_test.js
'use strict';

var Mt = require("./mt.js");
var Curry = require("../../lib/js/curry.js");
var Caml_format = require("../../lib/js/caml_format.js");
var Caml_exceptions = require("../../lib/js/caml_exceptions.js");
var Caml_js_exceptions = require("../../lib/js/caml_js_exceptions.js");

var suites = {
  contents: /* [] */0
};

var test_id = {
  contents: 0
};

function eq(loc, x, y) {
  return Mt.eq_suites(test_id, suites, loc, x, y);
}

var Inline_record = /* @__PURE__ */Caml_exceptions.create("Record_extension_test.Inline_record");

function f(x) {
  if (x.RE_EXN_ID === Inline_record) {
    return x.x + Caml_format.caml_int_of_string(x.y) | 0;
  }
  
}

var v0 = {
  RE_EXN_ID: Inline_record,
  x: 3,
  y: "4"
};

eq("File \"record_extension_test.ml\", line 19, characters 6-13", f(v0), 7);

function f2(x) {
  if (typeof x === "number" || x.TAG !== /* C */0) {
    return 0;
  } else {
    return x.x;
  }
}

function f2_with(x) {
  if (typeof x === "number" || x.TAG !== /* C */0) {
    return x;
  } else {
    return {
            TAG: /* C */0,
            x: 0,
            y: x.y
          };
  }
}

var A = /* @__PURE__ */Caml_exceptions.create("Record_extension_test.A");

var B = /* @__PURE__ */Caml_exceptions.create("Record_extension_test.B");

var C = /* @__PURE__ */Caml_exceptions.create("Record_extension_test.C");

function u(f) {
  try {
    return Curry._1(f, undefined);
  }
  catch (raw_x){
    var x = Caml_js_exceptions.internalToOCamlException(raw_x);
    if (x.RE_EXN_ID === A) {
      return x.name + x.x | 0;
    } else if (x.RE_EXN_ID === B) {
      return x._1 + x._2 | 0;
    } else if (x.RE_EXN_ID === C) {
      return x.name;
    } else {
      return -1;
    }
  }
}

Mt.from_pair_suites("File \"record_extension_test.ml\", line 56, characters 22-29", suites.contents);

exports.suites = suites;
exports.test_id = test_id;
exports.eq = eq;
exports.Inline_record = Inline_record;
exports.f = f;
exports.v0 = v0;
exports.f2 = f2;
exports.f2_with = f2_with;
exports.A = A;
exports.B = B;
exports.C = C;
exports.u = u;
/*  Not a pure module */

File: jscomp/test/record_extension_test.ml

(* for o in jscomp/test/*test.js ; do npx mocha  $o ; done *)

let suites :  Mt.pair_suites ref  = ref []
let test_id = ref 0
let eq loc x y = Mt.eq_suites ~test_id ~suites loc x y 


(* Record_extension *)
type t0  = ..
type t0 += Inline_record of { x : int ; y : string} 

let f x = 
  match x with 
  | Inline_record {x; y} -> Some (x + int_of_string y)  
  | _ -> None 
let v0 = Inline_record {x = 3; y = "4"}

;; eq __LOC__ (f v0) (Some 7)

(* Record_unboxed *)
type t1 = | A of { x : int } [@ocaml.unboxed]


(* Record_inlined *)
type t2 = 
  | B 
  | C of  {x : int ;y : string }
  | D of { w : int}
let f2 x = 
  match x with 
  | D _
  | B -> 0

  | C {x } -> x 

let f2_with x = 
  match x with 
  | D _
  | B -> x 
  | C u -> C {u with x = 0}


exception A of {name : int; x : int}  
exception B of int * int
exception C of {name : int }

let u f = 
  try f () with
  | A {name ; x} -> name + x 
  | B (a,b) -> a + b
  | C x -> x.name
  | _ -> -1

let () = 
  Mt.from_pair_suites __LOC__ !suites

